### PR TITLE
Simplify the creation of custom VMs in Crownlabs

### PR DESCRIPTION
**Pull request by ermete97**
# DESCRIPTION
This PR aims to simplify the creation of Custom VMs in Crownlabs using Packer.

### CODE REVIEW

File: provisioning/packer-virtual-machines/README.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+  - `ANSIBLE`, it indicates which ansible playbooks we want to run.
+- **[docker-image](docker-image/Dockerfile)**: as we can see in the Dockerfile, this image contains all it is necessary for login in the harbor registry, with the credential provided in the environment variables, and for creating and pushing the image in this registry. The environment variables are:
+  - `IMAGE_NAME`, it indicates the name of the image that will be pushed in Harbor;
+  - `USERNAME`, it indicates the username used for Harbor;
+  - `PASSWORD`, it indicates the password used for Harbor.
+
+## Creating and uploading a custom image
+For creating and uploading a custom image, we just have to run a job like [this](deploy/job.yaml). Clearly we have to change the environment variables in a properly way.
+This job use the two images saw [here](#images). In this case the packer-image is used as a `initContainer` and the docker-image is used as a simple container. In this way we can be sure that the uploading of the image in the Harbor registry happens after the creation of the image itself.
```
Comment (giorio94) [inline provisioning/packer-virtual-machines/README.md:28]:
```suggestion
It essentially combines the Docker images mentioned in the previous section. In this case the packer-image is used as an `initContainer` and the docker-image is used as a standard container. In this way we can be sure that the uploading of the image in the Harbor registry happens after the creation of the image itself.
```

File: provisioning/packer-virtual-machines/README.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+Once the job is started, we can completely disregard its execution. If everything goes well we can find our image in the repository indicated.
```
Comment (giorio94) [inline provisioning/packer-virtual-machines/README.md:29]:
```suggestion
The job executes completely unattended. Once it terminates successfully, the resulting image will be available in the configured registry.
```

File: provisioning/packer-virtual-machines/README.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+## Intro to Packer 
+[Packer](https://www.packer.io/) is an useful tool by Hashicorp for creating and configuring machine images in a very simple way.
+The `packer build` command takes a template and creates the artifact.
+It is based on three principal blocks:
+- **Builders**: they are blocks resposible for creating machines and generating images from them. There are a lot of Builders, for example for VirtualBox, VMware, etc. In our case we use QEMU Builder for creating a KVM virtual machine images.
+- **Provisioners**: they are responsible for installing and configuring the machine images after booting. For instance, thanks to Provisioners we can install packages and download application code. We use this block to run ansible playbooks and to clean cloud-init.
+- **Post-Processors**: they run after the image is built. With this block we can make some elaboration on the artifact created. We use this block for reducing the size of the image created through `virt-sparsify`.
+
+## Images
+Two images are involved in this mechanism:
```
Comment (giorio94) [inline provisioning/packer-virtual-machines/README.md:12]: Would clarify those are Docker images, and not VM images.

File: provisioning/packer-virtual-machines/deploy/job.yaml [hunk 4]
```diff
@@ -0,0 +1,60 @@
+apiVersion: batch/v1
```
Comment (giorio94) [inline provisioning/packer-virtual-machines/deploy/job.yaml:1]: Seems to me this is identical to the previous manifest, but for the `HAS_DESKTOP` variable. Would remove, and add the variable to the other. The more code/manifests are duplicated, the more inconsistencies you get.

File: provisioning/packer-virtual-machines/docker-image/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ENV USERNAME=
+ENV PASSWORD=
+
+CMD service docker start && \
+    rm -f Dockerfile && \
+    echo 'FROM scratch' >> Dockerfile && \
+    echo 'ADD output.qcow2 /disk/' >> Dockerfile && \
+    docker login -u "$(echo $USERNAME)" -p "$(echo $PASSWORD)" harbor.crownlabs.polito.it && \
+    docker build -t "$(echo $IMAGE_NAME)" . && \
+    docker push "$(echo $IMAGE_NAME)"
```
Comment (giorio94) [inline provisioning/packer-virtual-machines/docker-image/Dockerfile:28]: As we already discussed, this should be replace with [kaniko](https://github.com/GoogleContainerTools/kaniko), which is specifically thought to be executed as a container.

File: provisioning/packer-virtual-machines/packer-image/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+CMD if [ $BRANCH = master ]; then branch=trunk; else branch=branches/$(echo $BRANCH); fi && \
+    svn export --force $(echo $GIT_ANSIBLE_URL)/$(echo $branch)/provisioning/virtual-machines/ansible && \
+    packer build \
+    -force \
+    -var "ISO_URL=$(echo $ISO_URL)" \
+    -var "ISO_CHECKSUM=$(echo $ISO_CHECKSUM)" \
+    -var "ANSIBLE=$(echo $ANSIBLE)" \
+    -var "HAS_DESKTOP=$(echo $HAS_DESKTOP)" \
+    "builder.json"
```
Comment (giorio94) [inline provisioning/packer-virtual-machines/packer-image/Dockerfile:41]: It would be better to make this an entrypoint script to be executed.

File: provisioning/packer-virtual-machines/packer-image/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+ENV PACKER_LOG=1
+ENV ISO_URL=
+ENV ISO_CHECKSUM=
+ENV HAS_DESKTOP=false
+ENV GIT_ANSIBLE_URL=https://github.com/netgroup-polito/CrownLabs.git
+ENV BRANCH=master
+ENV ANSIBLE=
+
+CMD if [ $BRANCH = master ]; then branch=trunk; else branch=branches/$(echo $BRANCH); fi && \
+    svn export --force $(echo $GIT_ANSIBLE_URL)/$(echo $branch)/provisioning/virtual-machines/ansible && \
```
Comment (giorio94) [inline provisioning/packer-virtual-machines/packer-image/Dockerfile:34]: Why are you using svn instead of git?

File: provisioning/packer-virtual-machines/packer-image/packer_builder/builder.json [hunk 8]
```diff
@@ -0,0 +1,56 @@
+{
```
Comment (giorio94) [inline provisioning/packer-virtual-machines/packer-image/packer_builder/builder.json:1]: Be consistent. Since all other folders have dashes in their name, do not use underscores.

File: provisioning/packer-virtual-machines/packer-image/packer_builder/builder.json [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+            "iso_checksum": "{{ user `ISO_CHECKSUM` }}",
+            "cd_files": [
+                "./http/meta-data",
+                "./http/user-data"
+            ],
+            "cd_label": "cidata",
+            "output_directory": "output_ubuntu",
+            "shutdown_command": "echo 'crownlabs' | sudo -S shutdown -P now",
+            "memory": 2048,
+            "disk_size": "10G",
```
Comment (giorio94) [inline provisioning/packer-virtual-machines/packer-image/packer_builder/builder.json:16]: Should be made configurable.

File: provisioning/packer-virtual-machines/packer-image/packer_builder/builder.json [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+            "disk_compression": true,
+            "format": "qcow2",
+            "accelerator": "kvm",
+            "ssh_username": "ubuntu",
+            "ssh_password": "crownlabs",
+            "ssh_timeout": "30m",
```
Comment (giorio94) [inline provisioning/packer-virtual-machines/packer-image/packer_builder/builder.json:22]: Is this extra high timeout necessary?

File: provisioning/packer-virtual-machines/packer-image/packer_builder/builder.json [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+            "vm_name": "ubuntu.qcow2",
```
Comment (giorio94) [inline provisioning/packer-virtual-machines/packer-image/packer_builder/builder.json:23]:
```suggestion
            "vm_name": "crownlabs.qcow2",
```

File: provisioning/virtual-machines/ansible/roles/gather-desktop-environment-facts/tasks/main.yml [hunk 12]
```diff
@@ -6,20 +6,30 @@
...
+  when: desktop_mode is defined
 
 - name: Report whether a Desktop Environment has been detected
   debug:
     msg:
     - "Desktop Environment detected: {{ desktop_environment_detected }}"
-    - "XFCE detected: {{ xfce_detected }}"
\ No newline at end of file
+    - "XFCE detected: {{ xfce_detected }}"
+  when: desktop_mode is undefined
```
Comment (giorio94) [inline provisioning/virtual-machines/ansible/roles/gather-desktop-environment-facts/tasks/main.yml:35]: This change is not required.

File: provisioning/virtual-machines/ansible/roles/gather-desktop-environment-facts/tasks/main.yml [hunk 13]
```diff
@@ -6,20 +6,30 @@
   register: startx_presence_check
   changed_when: False
   failed_when: False
+  when: desktop_mode is undefined
```
Comment (giorio94) [inline provisioning/virtual-machines/ansible/roles/gather-desktop-environment-facts/tasks/main.yml:9]:
```suggestion
  when: install_desktop_environment is undefined or install_desktop_environment == "false"
```

Would be more explicit (same below)

File: provisioning/virtual-machines/ansible/roles/ubuntu-pre/tasks/main.yml [hunk 14]
```diff
@@ -38,6 +38,23 @@
...
+  dpkg_selections:
+    name: gdm3
+    selection: hold
+  when: desktop_mode is defined and desktop_mode == "true"
+
+- name: Install xubuntu-desktop
+  apt:
+    name: xubuntu-desktop
+    state: present
+  when: desktop_mode is defined and desktop_mode == "true"
```
Comment (giorio94) [inline provisioning/virtual-machines/ansible/roles/ubuntu-pre/tasks/main.yml:52]: Should pay attention, because `Remove many unnecessary packages` is removing many packages this step reinstalls. You should probably mark them to prevent their reinstallation.

File: provisioning/virtual-machines/ansible/roles/ubuntu-pre/tasks/main.yml [hunk 15]
```diff
@@ -38,6 +38,23 @@
...
+
+- name: Reboot the system
+  reboot:
+  when: desktop_mode is defined and desktop_mode == "true"
```
Comment (giorio94) [inline provisioning/virtual-machines/ansible/roles/ubuntu-pre/tasks/main.yml:56]: Why is this required?

**Review by giorio94 — COMMENTED** [review]
Some comments inline

**Comment by frisso** [comment]
Dear Ermete, I know that you're very busy now, but would it be possible to finalize this draft PR?
Many thanks.

**Comment by ermete97** [comment]
Yes, i'm working on it

File: provisioning/virtual-machines/ansible/roles/gather-desktop-environment-facts/tasks/main.yml (hunk 13 above)
Comment (giorio94) [reply provisioning/virtual-machines/ansible/roles/gather-desktop-environment-facts/tasks/main.yml:9]: Does not look resolved to me

**Review by giorio94 — APPROVED** [review]

File: provisioning/packer-virtual-machines/docker-image/Dockerfile (hunk 5 above)
Comment (giorio94) [reply provisioning/packer-virtual-machines/docker-image/Dockerfile]: Agreed to postpone to a follow-up PR
